File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0233_Number_of_Digit_One.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0233_Number_of_Digit_One {
    public int countDigitOne(int n) {
        long res = 0;
        for (long i = 1; i <= n; i *= 10) {
            long low = n % i;
            long cur = (n / i) % 10;
            long high = n / (i * 10);
            if (cur == 0) {
                res += high * i;
            } else if (cur == 1) {
                res += high * i + low + 1;
            } else {
                res += (high + 1) * i;
            }
        }
        return (int) res;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0233_Number_of_Digit_One.cc

class Solution {
 public:
  int countDigitOne(int n) {
    long long res = 0, high = 0, cur = 0, low = 0;
    for (long long i = 1; i <= n; i *= 10) {
      high = n / (i * 10);
      cur = (n / i) % 10;
      low = n % i;
      if (cur == 0) {
        res += high * i;
      } else if (cur == 1) {
        res += high * i + low + 1;
      } else {
        res += (high + 1) * i;
      }
    }
    return (int)res;
  }
};
